controlplane-rs/src/gateway_controller.rs

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use futures::StreamExt;
+use std::{
+    ops::Sub,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use crate::*;
+use k8s_gateway_api::{Gateway, GatewayClass, GatewayClassSpec, GatewaySpec};
+use k8s_openapi::api::core::v1::Service;
+use kube::{
+    api::{Api, ListParams, Patch, PatchParams},
+    runtime::{controller::Action, watcher::Config, Controller},
+    Resource, ResourceExt,
+};
+
+use chrono::Utc;
+use gateway_utils::*;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1 as metav1;
+use tracing::*;
+
+pub async fn reconcile(gateway: Arc<Gateway>, ctx: Arc<Context>) -> Result<Action> {
+    let start = Instant::now();
+    let client = ctx.client.clone();
+    let name = gateway.name_any();
+    let ns = gateway.namespace().unwrap();
+
+    let gateway_api: Api<Gateway> = Api::namespaced(client.clone(), &ns);
+    let gateway_spec: GatewaySpec = gateway.spec.clone();
+    let mut gw = Gateway {
+        metadata: gateway.metadata.clone(),
+        spec: gateway.spec.clone(),
+        status: gateway.status.clone(),
+    };
+
+    let gateway_class_api = Api::<GatewayClass>::all(client.clone());
+    let gateway_class = gateway_class_api
+        .get(gateway_spec.gateway_class_name.as_str())
+        .await
+        .map_err(Error::KubeError)?;
+
+    let gateway_class_spec: GatewayClassSpec = gateway_class.spec.clone();
```

Comment:
This stuff looks like setup that we might be able to re-use some between all the controllers, what do you think about creating some utility functions for this?

Reply:
not exactly sure how this can be re-used between controllers 😅 
everything here is statically typed, so can't think how we can have generic util functions for this